x-pack/plugins/ingest_manager/server/services/epm/packages/install.ts

```diff
@@ -95,6 +99,117 @@ export async function ensureInstalledPackage(options: {
+    if (installType === 'install' || installType === 'reinstall') {
+      logger.error(`uninstalling ${pkgkey} after error installing`);
+      await removeInstallation({ savedObjectsClient, pkgkey, callCluster });
+    }
+
+    if (installType === 'update') {
+      if (!installedPkg) {
+        logger.error(
+          `failed to rollback package after installation error ${error} because saved object was undefined`
+        );
+        return;
+      }
+      const prevVersion = `${pkgName}-${installedPkg.attributes.version}`;
+      logger.error(`rolling back to ${prevVersion} after error installing ${pkgkey}`);
+      await installPackage({
+        savedObjectsClient,
+        pkgkey: prevVersion,
+        callCluster,
+      });
+    }
+  } catch (e) {
+    logger.error(`failed to uninstall or rollback package after installation error ${e}`);
+  }
+}
+
+export async function upgradePackages({
+  savedObjectsClient,
+  packagesToUpgrade,
+  callCluster,
+}: {
+  savedObjectsClient: SavedObjectsClientContract;
+  packagesToUpgrade: string[];
+  callCluster: CallESAsCurrentUser;
+}): Promise<Array<UpgradePackageInfo | UpgradePackageError>> {
+  const res: Array<UpgradePackageInfo | UpgradePackageError> = [];
+
+  for (const pkgToUpgrade of packagesToUpgrade) {
+    let installedPkg: SavedObject<Installation> | undefined;
+    let latestPackage: RegistrySearchResult | undefined;
+    try {
+      [installedPkg, latestPackage] = await Promise.all([
+        getInstallationObject({ savedObjectsClient, pkgName: pkgToUpgrade }),
+        Registry.fetchFindLatestPackage(pkgToUpgrade),
+      ]);
+    } catch (e) {
+      res.push({ name: pkgToUpgrade, ...handleIngestError(e) });
+      continue;
+    }
+
+    // TODO I think we want version here and not `install_version`?
+    if (!installedPkg || semver.gt(latestPackage.version, installedPkg.attributes.version)) {
+      const pkgkey = Registry.pkgToPkgKey({
+        name: latestPackage.name,
+        version: latestPackage.version,
+      });
+
+      try {
+        const assets = await installPackage({ savedObjectsClient, pkgkey, callCluster });
+        res.push({
+          name: pkgToUpgrade,
+          newVersion: latestPackage.version,
+          oldVersion: installedPkg?.attributes.version ?? null,
+          assets,
+        });
+      } catch (e) {
+        res.push({ name: pkgToUpgrade, ...handleIngestError(e) });
+        await handleInstallPackageFailure(
+          savedObjectsClient,
+          e,
+          latestPackage.name,
+          latestPackage.version,
+          installedPkg,
+          callCluster
+        );
+      }
+    } else {
+      // package was already at the latest version
+      res.push({
+        name: pkgToUpgrade,
+        newVersion: latestPackage.version,
+        oldVersion: latestPackage.version,
+        assets: [
+          ...installedPkg.attributes.installed_es,
+          ...installedPkg.attributes.installed_kibana,
```

Comment:
In the case where no upgrade actually occurs because it was already at the latest version, should we return the current assets? I think this is all the installed assets?

Reply:
Good question.  Currently we reinstall the package if its already installed, but this functionality would not, so it would be a bit inconsistent, though I'm not sure we should be reinstalling anyway. I think for now its fine to just return the assets.